String.cxx

```diff
@@ -0,0 +1,184 @@
+    inline String(const char* cstr) : m_u{cstr} {}
+    inline explicit String(char c) : m_u{1, c} {}
+    template <class T, class = std::enable_if_t<std::is_integral<T>::value>> inline String(T val, StringBaseConv base) {
+        constexpr static auto base2len_upperbound = [](StringBaseConv base) /*constexpr*/ {
+            switch (base) {
+            case BIN:
+                return 65;
+            case DEC:
+
+            case HEX:
+                return 17;
+            default:
+                throw;
+            }
+        };
+        m_u.resize(base2len_upperbound(base));
+        const auto res = std::to_chars(&*m_u.begin(), &*m_u.rbegin(), val, +base);
+        if (static_cast<int>(res.ec))
+            throw;
+        m_u.resize(std::strlen(m_u.c_str()));
+    }
+
+     // template <class T, class = std::enable_if_t<std::is_floating_point<T>::value>>
+    // String(T val, int precision); // unimplemented
+
+    [[nodiscard]] inline auto c_str() const noexcept { return m_u.c_str(); }
+    [[nodiscard]] inline auto length() const noexcept { return m_u.length(); }
+    [[nodiscard]] inline auto charAt(unsigned idx) const noexcept { return m_u.at(idx); }
+    [[nodiscard]] inline auto& charAt(unsigned idx) noexcept { return m_u.at(idx); }
+    [[nodiscard]] inline auto operator[](unsigned idx) const noexcept { return m_u[idx]; }
+    [[nodiscard]] inline auto& operator[](unsigned idx) noexcept { return m_u[idx]; }
+
+    template <class T>[[nodiscard]] inline bool concat(const T& v) {
+        m_u += String(v).m_u;
+        return true;
+    }
+
+    inline int compareTo(const String& s) const noexcept { return std::memcmp(m_u.c_str(), s.m_u.c_str(), (std::min)(s.m_u.size(), m_u.size())); }
+
+    inline bool startsWith(const String& s) const noexcept {
+        if (s.m_u.size() > m_u.size())
+            return false;
+        return std::memcmp(m_u.c_str(), s.m_u.c_str(), s.m_u.size()) == 0;
+    }
+
+    inline bool endsWith(const String& s) const noexcept {
+        if (s.m_u.size() > m_u.size())
+            return false;
+        return std::memcmp(&*m_u.rbegin() - s.m_u.size(), s.m_u.c_str(), s.m_u.size()) == 0;
+    }
+
+    inline void getBytes(byte buffer [], unsigned length) { 
+        for (int i = 0; i < m_u.length; i++) {
+            buffer[i] = (byte)m_u[i];
+        }
+    }
+
+    inline int indexOf(const char* c) { return m_u.find(c); }
+
+    inline int indexOf(const char* c, unsigned index) { return m_u.find (c, index); }
+
+    inline int indexOf(const String& str) { return m_u.find(str); }
+
+    inline int indexOf(const String& str, unsigned index) { return m_u.find(str, index); }
+
+    inline void remove(unsigned idx) { m_u.erase(index); }
+
+    inline void remove(unsigned idx, unsigned count) { m_u.erase(idx, idx + count - 1); }
+
+    inline void replace(const String& substring1, const String& substring2) { m_u = std::regex_replace (m_u, std::regex(substring1), substring2); }
```

Comment:
I'm pretty sure regex is not wanted here.
[find](https://en.cppreference.com/w/cpp/string/basic_string/find) + [replace](https://en.cppreference.com/w/cpp/string/basic_string/replace) should be enough.

Reply:
`std::regex` is the slowest thing on earth; never use it

Reply:
Changed the implementation to find, replace